File: dotnet/Service/MongoResearcherService.cs
using BoilerplateCombo.Models;
using BoilerplateCombo.Repository;
using Cassandra;
using Microsoft.EntityFrameworkCore;
using MongoDB.Bson;
using MongoDB.Driver;

namespace BoilerplateCombo.Service;

public class MongoResearcherService(MongoClient client)
{
    public async Task<(List<Researcher2> researchers, int totalCount)> GetResearchersAsync(int page, int pageSize, string sortBy, bool ascending, string? filter)
    {
        var db = client.GetDatabase("bench");
        var col = db.GetCollection<BsonDocument>("researchers");
        var result = await col.FindAsync(new BsonDocument("researcher_id", "Alice"));
        return new();
    }

    public async Task<Researcher2?> GetResearcherByIdAsync(int id)
    {
        var db = client.GetDatabase("bench");
        var col = db.GetCollection<BsonDocument>("researchers");
        var result = await col.FindAsync(new BsonDocument("researcher_id", id));
        return null;
    }

    public async Task<Researcher2> AddResearcherAsync(Researcher2 researcher)
    {
        var db = client.GetDatabase("bench");
        var col = db.GetCollection<BsonDocument>("researchers");
        var doc = new BsonDocument { { "name", "Alice" }, { "age", 30 } };
        await col.InsertOneAsync(doc);
        return null;
    }

    public async Task<bool> UpdateResearcherAsync(int id, Researcher2 updatedResearcher)
    {
        var db = client.GetDatabase("bench");
        var col = db.GetCollection<BsonDocument>("researchers");
        var update = Builders<BsonDocument>.Update.Set("age", 31);
        await col.UpdateOneAsync(new BsonDocument("name", "Alice"), update);
        return true;
    }

    public async Task<bool> DeleteResearcherAsync(int id)
    {
        var db = client.GetDatabase("bench");
        var col = db.GetCollection<BsonDocument>("researchers");
        await col.DeleteOneAsync(new BsonDocument("name", "Alice"));
        return true;
    }

    public async Task<string> ExportResearchersAsCsvAsync()
    {
        // var researchers = await context.researcher.ToListAsync();
        // var csv = "Id,CreatedAt,Name\n";
        // foreach (var researcher in researchers)
        // {
        //     csv += $"{researcher.id},{researcher.created_at:O},{EscapeCsv(researcher.name)}\n";
        //     // $",{researcher.age}\n";
        // }
        // return 
        return "";
    }

    private string EscapeCsv(string value)
    {
        if (value.Contains(",") || value.Contains("\"") || value.Contains("\n"))
        {
            value = value.Replace("\"", "\"\"");
            return $"\"{value}\"";
        }
        return value;
    }
}

File: springboot/src/main/java/com/boilerplatecombo/service/MongoResearcherService.java
package com.boilerplatecombo.service;

import com.boilerplatecombo.entity.Researcher2;
import com.mongodb.client.result.DeleteResult;
import com.mongodb.client.result.UpdateResult;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.*;
import org.springframework.data.mongodb.core.query.*;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class MongoResearcherService {

    private final MongoTemplate mongoTemplate;

    public MongoResearcherService(@Qualifier("frosthandMongoTemplate") MongoTemplate mongoTemplate) {
        this.mongoTemplate = mongoTemplate;
    }

    public List<Researcher2> getResearchers(int page, int pageSize, String sortBy, boolean ascending, String filter) {
        Query query = new Query();

        if (filter != null && !filter.isEmpty()) {
            query.addCriteria(Criteria.where("name").regex(filter, "i"));
        }

        query.with(PageRequest.of(page, pageSize));
        query.with(Sort.by(ascending ? Sort.Direction.ASC : Sort.Direction.DESC, sortBy));

        return mongoTemplate.find(query, Researcher2.class);
    }

    public Researcher2 getResearcherById(String id) {
        return mongoTemplate.findById(id, Researcher2.class);
    }

    public Researcher2 addResearcher(Researcher2 researcher) {
        return mongoTemplate.insert(researcher);
    }

    public boolean updateResearcher(Long id, Researcher2 updatedResearcher) {
        Query query = new Query(Criteria.where("_id").is(id));
        Update update = new Update()
                .set("name", updatedResearcher.getName())
                .set("age", updatedResearcher.getAge());
        UpdateResult result = mongoTemplate.updateFirst(query, update, Researcher2.class);
        return result.getModifiedCount() > 0;
    }

    public boolean deleteResearcher(Long id) {
        Query query = new Query(Criteria.where("_id").is(id));
        DeleteResult result = mongoTemplate.remove(query, Researcher2.class);
        return result.getDeletedCount() > 0;
    }

    public String exportResearchersAsCsv() {
        List<Researcher2> researchers = mongoTemplate.findAll(Researcher2.class);
        StringBuilder csv = new StringBuilder("Id,Name,Age\n");
        for (Researcher2 r : researchers) {
            csv.append(String.format("%s,%s,%d\n",
                    escapeCsv(r.getId().toString()), escapeCsv(r.getName()), r.getAge()));
        }
        return csv.toString();
    }

    private String escapeCsv(String value) {
        if (value == null) return "";
        if (value.contains(",") || value.contains("\"") || value.contains("\n")) {
            return "\"" + value.replace("\"", "\"\"") + "\"";
        }
        return value;
    }
}
